crates/fetch_queue/src/fetch_queue.rs

```diff
@@ -0,0 +1,131 @@
+use std::collections::HashSet;
+
+use kitsune2_api::{fetch::FetchQueueT, AgentId, OpId};
+use rand::seq::IteratorRandom;
+
+#[derive(Debug)]
+pub struct FetchQueue {
+    ops: Vec<OpId>,
```

Comment:
These fields need to be grouped in some way. Not every agent is a valid source for every op. In a sharded network, that will certainly not be the case. Even unsharded, we can't assume that every agent have a given op unless they've told us they do because it might be a new agent who is catching up

Reply:
Also, we need to accommodate the request for validation receipts flags thing. Maybe something like:

```rust
type Flags = u32;
struct Q(VecDeque<(OpId, HashSet<AgentId>, Flags)>);
```

The `VecDeque` gives it actual queue properties, but perhaps on fresh syncs will contain too many items to be efficient enough? We could bring in IndexMap to keep them in order:

```rust
type Flags = u32;
struct Q(IndexMap<OpId, (HashSet<AgentId>, Flags)>);
```

Reply:
@ThetaSinner I got hung up on your idea of not fetching ops immediately from the same source we got the op list from, and randomizing who we fetch from. In the process of writing the code, I also thought it's very likely that not all agents will have the same range of ops to respond with. That's actually what I meant in Mattermost when I asked what to do when agents don't have all ops, and if we wanted to treat that optimistically and assume, it'll be complete eventually.

Too much thinking to myself and too little context in my messages. I agree this needs to be per agent, and now we're on the same page regarding immediate fetches as well.

Reply:
@neonphog IndexMap is used in kitsune1 too, I like that.

Reply:
What do you think of indexing by agent id instead of op id?
```rust
struct Q(IndexMap<AgentId, (Vec<OpId>, Vec<Flags>)>);
```

A fetch request takes `max_hash_count` number of elements from the vector of the first of the map's values, until that vector is empty. Then the next agent id's op list is processed.

Otherwise with the op id as index, ops are added individually with the agent as source, and when making a fetch request, you'd pop the first op id and then collect `max_hash_count` of op ids that are also in the map for that source. Seems more complex.

@neonphog @ThetaSinner

Reply:
There are 2 concerns here:

- First, we want the ops roughly in the order in which they were requested. Because a fresh sync will be going by time older to newer. This feels like a pretty soft requirement, because, frankly, who cares. We eventually need to get all of them, do we even actually care about the order?
- Second, it makes sense to optimize a request to a particular peer... that is, send a request for all the ops we think that peer will be able to respond with. This seems like a harder requirement because there is real-world overhead to connection establishment and bandwidth usage.

The downside of doing it the way you suggest is the dis-correlation of OpIds. The op data is the real bandwidth bottleneck, so I suggest we do not want to be making parallel requests for the op data at the same time. And this suggested data structure will make it very difficult to prevent this.

Reply:
I thought we did care about the order, but this is a part that leaves room for a lot different approaches and speculations.

Got your second point, that is a clear no to this other data structure. Another downside is that redundant op fetches are far more likely or need to be tracked in some way.